[jit] `__copy__` for `RecursiveScriptModule`

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32685 [jit] Remove `copy` from public API
* #37538 [jit][api][refactor] remove redundant deepcopy implementation
* **#36830 [jit] `__copy__` for `RecursiveScriptModule`**

Summary:

Test Plan:
build/bin/test_jit

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D21431012](https://our.internmc.facebook.com/intern/diff/D21431012)

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -470,6 +470,57 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
+    case IValue::Tag::Tuple: {
+      std::vector<IValue> copied_tuple;
+      for (const auto& e : toTuple()->elements()) {
+        copied_tuple.push_back(e.copy());
```
A shallow copy shouldn't be copying the subreferences right?

jerryzh168 replied:
I see, will change

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -470,6 +470,57 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
+      auto list = toList();
+      auto copied_list = c10::impl::GenericList(list.elementType());
+      for (IValue v : list) {
+        copied_list.push_back(v.copy());
```
same here

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -470,6 +470,57 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
+      auto dict = toGenericDict();
+      auto copied_dict = c10::impl::GenericDict(dict.keyType(), dict.valueType());
+      for (const auto& entry : dict) {
+        copied_dict.insert(entry.key().copy(), entry.value().copy());
```
and here

zdevito reviewed (commented):
Is this a shallow copy? It seems to be doing a deep copy without memoization? The behavior of the is function should be aligned with python.

zdevito commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -470,6 +470,39 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
+  IValue copy;
+  switch(tag) {
+    case IValue::Tag::Tensor:
+      copy = IValue(toTensor());
```
These statements are not copying these objects, they are just doing a round trip IValue -> Reference -> IValue

jerryzh168 replied:
Yeah, this is meant to be a shallow copy, so only the IValue wrapper is updated

zdevito commented on torch/csrc/jit/api/object.cpp:
```diff
@@ -35,6 +35,18 @@ void Object::define(const std::string& src, const ResolverPtr& resolver) {
 }
 
+Object Object::copy() const {
+  Object obj(_ivalue()->compilation_unit(), type());
```
This code is duplicated with `ivalue::Object::copy`, and should just call it instead.

zdevito commented on torch/csrc/jit/api/object.cpp:
```diff
@@ -35,6 +35,18 @@ void Object::define(const std::string& src, const ResolverPtr& resolver) {
+  size_t N = type()->numAttributes();
+  for (size_t i = 0; i < N; ++i) {
+    IValue s = _ivalue()->getSlot(i);
+    obj._ivalue()->setAttr(type()->getAttributeName(i), s.copy());
```
calling `s.copy()` makes it a deep copy

jerryzh168 replied:
I see, so we just need to copy the top level wrapper for shallow copy?

zdevito reviewed (commented):
Thanks for updating this. There are still a number of bugs making it not consistent whether this is a deep or shallow copy. The fact that there are no tests to catch these bugs is concerning as well.

zdevito reviewed (commented):
The current code is not copying (even shallow) anything. Object and Module are just wrapper classes around `ivalue`. A shallow copy of an object needs to actually create a copy of the underlying ivalue.

jerryzh168 commented:
> The current code is not copying (even shallow) anything. Object and Module are just wrapper classes around `ivalue`. A shallow copy of an object needs to actually create a copy of the underlying ivalue.

so only create a new copy of the underlying ivalue::Object, but don't copy the `c10::IValue` inside the `ivalue::Object`?

jerryzh168 committed:
Update on "[jit] `__copy__` for `RecursiveScriptModule`"


Summary:

Test Plan:
build/bin/test_jit

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[jit] `__copy__` for `RecursiveScriptModule`"


Summary:

Test Plan:
build/bin/test_jit

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@zdevito could you review again?

zdevito reviewed (approved):
Thanks!

jerryzh168 committed:
Update on "[jit] `__copy__` for `RecursiveScriptModule`"


Summary:

Test Plan:
build/bin/test_jit

Reviewers:
.

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
